libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,333 @@
+        CofactorExists (minus qab qa) positB (
+            rewrite multDistributesOverMinusRight p qab qa in
+            rewrite prfA in
+            rewrite prfAB in
+            sym (
+                (b) ={ sym $ minusZeroRight b }=
+                (minus b 0) ={ sym $ plusMinusLeftCancel a b 0 }=
+                (minus (a + b) (a + 0)) ={ replace {P = \x => minus (a + b) (a + 0) = minus (a + b) x} (plusZeroRightNeutral a) Refl }=
+                (minus (a + b) a) QED
+            )
+        )
+
+||| If p is a common factor of a and b, then it is also a factor of their GCD.
+||| This actually follows directly from the definition of GCD.
+commonFactorAlsoFactorOfGCD : Factor a p -> Factor b p -> GCD a b q -> Factor q p
+commonFactorAlsoFactorOfGCD {p} pfa pfb (MkGCD _ greatest) =
+        greatest p (CommonFactorExists p pfa pfb)
+
+
+||| A decision procedure for whether of not p is a factor of n.
+decFactor : (n, d : Nat) -> {auto nok : LTE 1 n} -> {auto dok : LTE 1 d} -> DecFactor n d
+decFactor n (S d) {nok} {dok} with (Data.Fin.Extra.divMod n (S d))
+        | (Fraction n (S d) q r prf) = case r of
+                FZ =>
+                    let prf =
+                            replace {P = \x => x = n} (plusZeroRightNeutral (q + (d * q))) $
+                            replace {P = \x => x + 0 = n} (multCommutative q (S d)) prf
+                    in
+                    ItIsFactor $ CofactorExists q nok prf
+
+                (FS pr) =>
+                    ItIsNotFactor $ ProperRemExists q pr nok (
+                            rewrite multCommutative d q in
+                            rewrite sym $ multRightSuccPlus q d in
+                            prf
+                        )
+
+||| For all p greater than 1, if p is a factor of n, then it is NOT a factor
+||| of (n + 1).
+factNotSuccFact : {n, p : Nat} -> GT p 1 -> Factor n p -> NotFactor (S n) p
+factNotSuccFact {n} {p = Z} pGt1 (CofactorExists q positN prf) =
+        absurd $ succNotLTEzero pGt1
+factNotSuccFact {n} {p = S Z} pGt1 (CofactorExists q positN prf) =
+        absurd . succNotLTEzero $ fromLteSucc pGt1
+factNotSuccFact {n} {p = S (S k)} pGt1 (CofactorExists q positN prf) =
+        let r = FZ in -- remember it's remainders precedessor
+        ProperRemExists q r (lteSuccRight positN) (
+            rewrite prf in
+            rewrite plusCommutative n 1 in
+            Refl
+        )
+
+||| 1 is a common factor of any pair of natural numbers.
+oneCommonFactor : (a, b : Nat) -> {auto aok : LTE 1 a} -> {auto bok : LTE 1 b} -> CommonFactor a b 1
+oneCommonFactor a b {aok} {bok} = CommonFactorExists 1
+        (CofactorExists a aok (rewrite plusZeroRightNeutral a in Refl))
+        (CofactorExists b bok (rewrite plusZeroRightNeutral b in Refl))
+
+||| Any natural number is a common factor of itself and itself.
+selfIsCommonFactor : (a : Nat) -> {auto ok : LTE 1 a} -> CommonFactor a a a
```

Comment:
Same here, I would call this `commonFactorReflexive` and move it next to `commonFactorSym`. Is there a sense in which `CommonFactor` is transitive? If it is, you could implement it as an [equivalence relation](https://github.com/idris-lang/Idris-dev/blob/0417c53fb6b91fc0692245f269b98a3bd4820136/libs/contrib/Decidable/Order.idr#L29).

Reply:
Well, in order to be transitive, it would have to be a two-argument predicate. This can be done like so:

```
CommonFactor (a, b) p
```
However now sides of the relation have different types, which is no good. However, if you stretch the definition a little bit, you can come up with something like this:

```
cfTransitive : CommonFactor (a, b) p -> CommonFactor (b, c) p -> CommonFactor (a, c) p
```
So it is transitive when you consider it a binary relation (either there is or there isn't a common factor of two numbers in question), without considering what that factor actually is. When looked upon like this, it is indeed transitive, and this is the sense in which it is symmetric in `commonFactorSym` proof.